compiler/riscv/codegen/BinaryEvaluator.cpp

```diff
@@ -295,37 +330,324 @@ OMR::RV::TreeEvaluator::ludivEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    return Rhelper(node, TR::InstOpCode::_divu, cg);
    }
 
-
 TR::Register *
 OMR::RV::TreeEvaluator::lremEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
    return Rhelper(node, TR::InstOpCode::_rem, cg);
    }
 
-static TR::Register *shiftHelper(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::InstOpCode::Mnemonic opi, TR::CodeGenerator *cg)
-   {
-   // TODO: if second arg is an int constant, shall we make mask it so
-   // that we're sure the only low 5 bits are set? This is the assumption
-   // under which we can use RorIhelper. Is this guaranteed at this level?
-   return RorIhelper(node, op, opi, cg);
-   }
-
+// also handles bshl, sshl and lshl
 TR::Register *
 OMR::RV::TreeEvaluator::ishlEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return shiftHelper(node, TR::InstOpCode::_sllw, TR::InstOpCode::_slliw, cg);
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+
+   TR::Register *src1Reg;
+   TR::Register *trgReg;
+
+   int width = TR::DataType::getSize(node->getDataType()) * 8;
+
+   if (secondChild->getOpCode().isLoadConst())
+      {
+      /*
+       * If shift amount is a constant value, prefer slli / slliw.
+       *
+       * Only low bits 6 bits (5 for Int32, 4 for Int16, 3 for Int8)
+       * are considered, higher bits are ignored. This is to be consistent
+       * with behavior of sll / sllw.
+       *
+       * If data width is 8 or 16 bits, we have to truncate the value.
+       * We do this by combining the requested left shift by a constant
+       * with "truncation" left shift, for example:
+       *
+       *   For...
+       *
+       *     (sshl
+       *        (<src1>)
+       *        (iconst 5)
+       *
+       *   ...we generate:
+       *
+       *       slliw trg, src1, 21   ;; 21 = 32(reg width) - 16(data width) + 5(requested shift amount)
+       *       sraiw trg, trg,  16   ;; 16 = 32(reg width) - 16(data width)
+       *
+       *   This saves us one instruction compared to doing
+       *   `slliw trg, src1, 5` and then standard truncate()
+       *
+       * If constant is 0, we just pass-through the source value.
+       */
+      int64_t shamt = secondChild->getLongInt() & (width - 1);
+      if (shamt != 0)
+         {
+         if (width < 32)
+            {
+            if (cg->canClobberNodesRegister(firstChild))
+               {
+               src1Reg = trgReg = cg->gprClobberEvaluate(firstChild);
+               }
+            else
+               {
+               src1Reg = cg->evaluate(firstChild);
+               trgReg = cg->allocateRegister(TR_GPR);
+               }
```

Comment:
Is there a particular reason to try and reuse `firstChild`'s register? Usually we would try to reuse child registers to avoid unnecessary register copies or to reduce register pressure of internal control flow, but it doesn't look like either of those are happening in these evaluators.

Reply:
You're right, there's no reason for that. I'll change it.